### DISCUSSION TITLE
Publish Message .NetCore

### DISCUSSION THREAD

**Discussion by Mauro8k** [rebus-org/Rebus#1096] [Q&A] [answered]
Hi 
I am trying to send a meesage to rabbitMq from a  .NetCore API (POST Method), I made all the configuration : 
Install the Nyget package
* Installed the Nuget Rebus.RabbitMq 8.0.0
* nstalled the Nuget rebus.ServiceProvider 9.0.0
 I configured . Starup file , and created the all logic to send the message.

But this is the erro the RabbitMq show me : WebApp could not be dispatched to any handlers (and will not be retried under the default fail-fast settings)) rebus.

the most curios thing is when I tested the same code logic con configurations but using the Nuget rebus.ServiceProvider 7.0.0
works fine. Hope you can hepl me with the issue. 
Thanks in advance.

**Comment by mookid8000** [comment]
The error message

> <some message type> could not be dispatched to any handlers (and will not be retried (...))

means that a message has been received from a queue, but Rebus was unable to find an appropriate handler for the message.

EITHER you're missing a handler registration, e.g. like

```csharp
services.AddRebusHandler<SomeMessageHandler>();
```

OR you have sent the message to the wrong queue.

If you tell me exactly how you sent the message (e.g. was it `await bus.Send(..)`, `await bus.Publish(..)`?) and which queues and Rebus instances you have, I can help you diagnose the issue. 🙂

**Reply by Mauro8k** [reply]
II am using await _bus.Send("message");
but I have curious because I have a console app working as handler, even is not working , so my question is should I have the handler on the same API project or still working the handler on appconsole.
Thank you for your response.

**Comment by mookid8000** [comment, marked as answer]
When you `await _bus.Send("message");` you must have configured a destination for messages of type `System.String` like this:

```csharp
services.AddRebus(
    configure => configure
        .(...)
        .Routing(t => t.TypeBased().Map<string>("some-queue"))
);
```

which will cause the message to be sent to the queue named "some-queue". Now you'll just have to be sure that the Rebus instance that has "some-queue" as its input queue has an appropriate message handler registered, e.g. like this:

```csharp
services.AddRebusHandler<StringHandler>();
```

where `StringHandler` could be something like this:

```csharp
public class StringHandler : IHandleMessages<string>
{
    public async Task Handle(string message)
    {
        // .. handle message in here
    }
}
```

**Reply by Mauro8k** [reply]
Yes I have the same configurations on my API project , but I did not have the  services.AddRebusHandler<StringHandler>(); on my API I created one and worked well thank you so much :)

**Comment by mookid8000** [comment]
Ok, cool! 🙂 But just to be sure (because this is a mistake that I've often seen people make when they start out with Rebus): Your two projects each have their own input queue, right?

Because if you let two Rebus instances share the same input queue, only one of them will get each message, and it will result in the same error message from time to time.

**Reply by Mauro8k** [reply]
just to make sure the scenario is : 

I have a .net core API to send a message into  queue1 and I have a other project a  console app the wants to read the message from the   queue1, but i am afraid the app console handler is not working , the handle only works I have the handler in the same project , so am I missing some handler configuration on the app console?

**Reply by mookid8000** [reply]
If your .NET Core API sends the message to "queue1", and your console app has "queue1" as its input queue, then everything should work.

But please note that it should ONLY be your console app that has "queue1" as its input queue.

**Reply by Mauro8k** [reply]
this is the API configuration

builder.Services.AddRebus(configure => configure
           .Transport(x => x.UseRabbitMq("amqp://guest:guest@localhost:5672", "Bus.Users"))
           .Routing(x => x.TypeBased()
                .Map<UserCreatedEvent>("Bus.Users"))
           );

sending the message to Bus.Users" queue

this the config for console app (handler)

using var activator = new BuiltinHandlerActivator();
activator.Register(() => new UserCreatedEventHandler());
var subscriber = Configure.With(activator)
    .Transport(t => t.UseRabbitMq("amqp://guest:guest@localhost:5672", "Bus.Users"))
    .Start();
await subscriber.Subscribe<UserCreatedEvent>();

Handler
public class UserCreatedEventHandler : IHandleMessages<UserCreatedEvent>
    {
        public async Task Handle(UserCreatedEvent message)
        {
            Console.WriteLine($"{nameof(UserCreatedEvent)} received. Username: {message.UserName}");
        }
    }

**Reply by mookid8000** [reply]
Ok that will not work.

You have two Rebus instances using the input queue "Bus.Users", which means that they will compete for receiving messages. Each message will end up being consumed either by your API or by your console app.

If you change your API configuration to

```csharp
builder.Services.AddRebus(configure => configure
    .Transport(x => x.UseRabbitMq("amqp://guest:guest@localhost:5672", "MyWebApi"))
    .Routing(x => x.TypeBased().Map<string>("Bus.Users"))
);
```

then it will work. Or you may even want to consider making your API's Rebus instance a "one-way client" (i.e. a Rebus instance only capable of sending/publishing) like this:

```csharp
builder.Services.AddRebus(configure => configure
    .Transport(x => x.UseRabbitMqAsOneWayClient("amqp://guest:guest@localhost:5672"))
    .Routing(x => x.TypeBased().Map<string>("Bus.Users"))
);
```
